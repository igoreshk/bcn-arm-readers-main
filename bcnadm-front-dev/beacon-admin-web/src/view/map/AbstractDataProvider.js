/* eslint-disable no-unused-vars */
/* @flow*/

class AbstractDataProvider {
  /**
   * @param {string} buildingId
   * @param {string} levelId
   * @param {object} dto
   * @param {object} entityId
   * @returns {Promise}
   */
  getDtos(buildingId, levelId) {
    throw Error('Method not supported');
  }

  getDto(buildingId, levelId, markerId) {
    throw Error('Method not supported');
  }

  saveDto(buildingId, levelId, dto) {
    throw Error('Method not supported');
  }

  removeDto(buildingId, levelId, dto) {
    throw Error('Method not supported');
  }

  removeAll(buildingId, levelId) {
    throw Error('Method not supported');
  }
}

export default AbstractDataProvider;
