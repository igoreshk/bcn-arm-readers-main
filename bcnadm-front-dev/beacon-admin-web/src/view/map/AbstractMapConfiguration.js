/* eslint-disable no-unused-vars */
/* @flow*/

export default class AbstractMapConfiguration {
  /**
   * createMap
   * @returns {Promise}
   */
  createMap() {
    throw Error('Method not supported');
  }

  /**
   * setMapBackground
   * @param {object} map
   * @returns {void}
   */
  setMapBackground(map) {
    throw Error('Method not supported');
  }

  /**
   * getImageAndCalculateBounds
   * @param {object} map
   * @param {string} levelId
   * @param {string} buildingId
   * @returns {Promise}
   */
  getImageAndCalculateBounds(map, levelId, buildingId) {
    throw Error('Method not supported');
  }

  /**
   * configureOverlay
   * @param {object} map
   * @param {object} bounds
   * @param {string} image
   * @returns {Promise}
   */
  configureOverlay(map, bounds, image) {
    throw Error('Method not supported');
  }

  /**
   * setMapDraggingListener
   * @param {object} map
   * @param {object} bounds
   */
  setMapDraggingListener(map, bounds) {
    throw Error('Method not supported');
  }
}
