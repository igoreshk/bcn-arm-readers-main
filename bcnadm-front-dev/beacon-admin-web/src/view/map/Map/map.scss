.mapWrapper {
  flex: 1 0;
  height: calc(100vh - 56px);
  z-index: 1;
}
.mapContainer {
  height: 100%;
}
