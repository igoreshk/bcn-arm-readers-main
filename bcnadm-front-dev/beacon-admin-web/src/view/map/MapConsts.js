/**
 * Contains set of constants for configuring map
 * @namespace mapConsts
 * @prop {object} mapOptions Constant, used in google.maps.Map() constructor when creating map
 * @prop {function} configureCoordMapType Configures overlay map type & map tiles
 * (See: {@link configureCoordMapType})
 * @prop {function} configureOverlay Configures overlay, particularly sets custom image as an overlay, size of an overlay etc
 * (See: {@link configureOverlay})
 *
 * @see [Custom overlays]{@link https://developers.google.com/maps/documentation/javascript/customoverlays?hl=en}
 * To read about configuring custom overlays
 */
export const INITIAL_ZOOM = 18;

const mapConsts = {
  mapOptions: {
    zoom: INITIAL_ZOOM,
    minZoom: 12,
    maxZoom: 28,
    center: { lat: 0, lng: 0 },
    mapTypeId: 'roadmap',
    zoomControl: true,
    mapTypeControl: false,
    scaleControl: true,
    streetViewControl: false,
    rotateControl: false,
    fullscreenControl: false
  },
  /**
   * This function configures tiles of map
   * @function
   * @param coord {object} Coordinates on current tile
   * @param zoom {number} Zoom rate
   * @param ownerDocument {object} returns higher-level document object for current node
   * @returns {object} configured HTML element
   */
  divConfig(coord, zoom, ownerDocument) {
    const div = ownerDocument.createElement('div');
    div.innerHTML = coord;
    div.style.width = `${this.tileSize.width}px`;
    div.style.height = `${this.tileSize.height}px`;
    div.style.fontSize = '0';
    div.style.borderStyle = 'solid';
    div.style.borderWidth = '10000px';
    div.style.borderColor = '#e5e5e5';
    return div;
  },
  /**
   * This function is called when the map is ready for the overlay to be attached
   *@function
   */
  onAdd() {
    const div = document.createElement('div');
    div.style.border = 'none';
    div.style.borderWidth = '0px';
    div.style.position = 'absolute';

    const img = document.createElement('img');
    img.src = this.image;
    img.style.width = '100%';
    img.style.height = '100%';
    div.appendChild(img);

    this.div = div;
    const panes = this.getPanes();
    panes.overlayLayer.appendChild(div);
  },
  /**
   * This function  will be called when the object (custom overlay) is first displayed
   * @function
   */
  draw() {
    const overlayProjection = this.getProjection();
    const sw = overlayProjection.fromLatLngToDivPixel(this.bounds.getSouthWest());
    const ne = overlayProjection.fromLatLngToDivPixel(this.bounds.getNorthEast());
    const div = this.div;
    this.div.id = 'layer';
    div.style.left = `${sw.x}px`;
    div.style.top = `${ne.y}px`;
    div.style.width = `${ne.x - sw.x}px`;
    div.style.height = `${sw.y - ne.y}px`;
    div.style.transform = `rotate(${this.rotation}rad)`;
  },
  /**
   * This function is used to clean up any elements added within the overlay.
   * @function
   */
  onRemove() {
    this.div.parentNode.removeChild(this.div);
    this.div = null;
  }
};
export const TILE_SIZE = 1000;
export default mapConsts;
/**
 * Mode when user adds beacon markers
 * @constant
 * @type {string}
 */
const BEACON_MODE = 'beacons';
/**
 * Mode when user adds reader markers
 * @constant
 * @type {string}
 */
const READER_MODE = 'readers';
/**
 * Mode when user adds vertex markers
 * @constant
 * @type {string}
 */
const ROUTE_MODE = 'routing';
/**
 * Mode when user adds area markers
 * @constant
 * @type {string}
 */
const AREAS_MODE = 'areas';
/**
 * Mode when user monitore people on the map
 * @constant
 * @type {string}
 */
const MONITORING_MODE = 'monitoring';
/**
 * Mode when user is able to draw on the map tile
 * @constant
 * @type {string}
 */
const DRAWING_MODE = 'drawing';
/**
 * Represents state of a dialog, where beacon's info can be configures
 * @constant
 * @type {string}
 */
const BEACON_DIALOG_OPEN = 'BEACON_DIALOG_OPEN';
/**
 * Represents state of a dialog, where beacon's info can be configures
 * @constant
 * @type {string}
 */
const BEACON_DIALOG_CLOSE = 'BEACON_DIALOG_CLOSE';
/**
 * Represents state of a dialog, where area's info can be configures
 * @constant
 * @type {string}
 */
const AREAS_DIALOG_OPEN = 'AREAS_DIALOG_OPEN';

const READER_DIALOG_OPEN = 'READER_DIALOG_OPEN';

const READER_DIALOG_CLOSE = 'READER_DIALOG_CLOSE';

const DISTANCE_DIALOG_OPEN = 'DISTANCE_DIALOG_OPEN';

const DISTANCE_DIALOG_CLOSE = 'DISTANCE_DIALOG_CLOSE';

/**
 * Represents state of a dialog, where area's info can be configures
 * @constant
 * @type {string}
 */
const AREAS_DIALOG_CLOSE = 'AREAS_DIALOG_CLOSE';

/**
 * Describes different modes
 * @type {{BEACON_MODE: string, ROUTE_MODE: string, AREAS_MODE: string}}
 */
const SCALE_MODE = 'scaling';

const GOOGLE = 'Google';

const OSM = 'OSM';

export const modes = {
  READER_MODE,
  BEACON_MODE,
  ROUTE_MODE,
  AREAS_MODE,
  SCALE_MODE,
  MONITORING_MODE,
  DRAWING_MODE
};
/**
 * Set of area/beacon dialog states
 * @constant
 * @type {object}
 */
export const dialog = {
  BEACON_DIALOG_OPEN,
  BEACON_DIALOG_CLOSE,
  AREAS_DIALOG_OPEN,
  AREAS_DIALOG_CLOSE,
  DISTANCE_DIALOG_OPEN,
  DISTANCE_DIALOG_CLOSE,
  READER_DIALOG_OPEN,
  READER_DIALOG_CLOSE
};

export const mapProviders = {
  GOOGLE,
  OSM
};
