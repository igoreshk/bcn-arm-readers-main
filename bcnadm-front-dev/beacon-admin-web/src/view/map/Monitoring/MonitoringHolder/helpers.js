export const updateComponentIfPropsChanged = (props, nextProps) => {
  // for re-render after loading the map
  if (!props.selectedMap && nextProps.selectedMap) {
    return true;
  }

  // For re-render after map changes
  if (props.map !== nextProps.map) {
    return true;
  }

  // For re-render after mapProvider changes
  if (props.mapProvider !== nextProps.mapProvider) {
    return true;
  }

  // For re-render after selectedMap changes
  if (props.selectedMap !== nextProps.selectedMap) {
    return true;
  }

  // For re-render after choosing another floor
  if (props.levelNumber !== nextProps.levelNumber) {
    return true;
  }

  return false;
};
