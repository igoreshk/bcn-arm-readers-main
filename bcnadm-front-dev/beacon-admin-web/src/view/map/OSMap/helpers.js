import L from 'leaflet';

export const calculatePixelsPerMeter = (map, level) => {
  const { scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude, scaleDistance } = level;

  const baseValue = 10;
  if (scaleDistance === 0) {
    return baseValue;
  }
  const latLngPoints = [
    L.latLng(scaleStartLatitude, scaleStartLongitude),
    L.latLng(scaleEndLatitude, scaleEndLongitude)
  ];
  const points = [map.latLngToLayerPoint(latLngPoints[0]), map.latLngToLayerPoint(latLngPoints[1])];
  const degree = 2;
  const distanceInPixels = Math.sqrt((points[0].x - points[1].x) ** degree + (points[0].y - points[1].y) ** degree);
  return distanceInPixels / scaleDistance;
};

export const rotateImage = (map, level) => {
  const { scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude } = level;

  if ((scaleStartLatitude === 0 && scaleStartLongitude === 0) || (scaleEndLatitude === 0 && scaleEndLongitude === 0)) {
    return 0;
  }

  const A = map.project(L.latLng(scaleStartLatitude, scaleStartLongitude));
  const B = map.project(L.latLng(scaleEndLatitude, scaleEndLongitude));
  const degree = 2;
  const hypotenuse = Math.sqrt((A.x - B.x) ** degree + (A.y - B.y) ** degree);
  const cathetus = Math.abs(A.y - B.y);
  let alpha = Math.acos(cathetus / hypotenuse);
  if ((A.x > B.x && A.y > B.y) || (A.x < B.x && A.y < B.y)) {
    alpha = Math.PI - alpha;
  }
  return alpha;
};
