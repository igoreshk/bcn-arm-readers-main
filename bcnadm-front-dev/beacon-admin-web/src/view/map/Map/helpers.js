import { INITIAL_ZOOM } from 'view/map/MapConsts';

export const isScaleEdgeChanged = (prevLevel, currentLevel) => {
  return (
    prevLevel.scaleStartLatitude !== currentLevel.scaleStartLatitude ||
    prevLevel.scaleStartLongitude !== currentLevel.scaleStartLongitude ||
    prevLevel.scaleEndLatitude !== currentLevel.scaleEndLatitude ||
    prevLevel.scaleEndLongitude !== currentLevel.scaleEndLongitude ||
    prevLevel.scaleDistance !== currentLevel.scaleDistance
  );
};

// Toggle background if map is not null
export const toggleBackgroundIfMapExist = (map, toggleBackgroundFunc) => {
  map && toggleBackgroundFunc();
};

export const calculatePixelsPerMeter = (map, google, level) => {
  const { scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude, scaleDistance } = level;

  const baseValue = 10;
  if (scaleDistance === 0) {
    return baseValue;
  }

  const degree = 2;
  const scale = degree ** INITIAL_ZOOM;
  const points = [
    map.getProjection().fromLatLngToPoint(new google.maps.LatLng(scaleStartLatitude, scaleStartLongitude)),
    map.getProjection().fromLatLngToPoint(new google.maps.LatLng(scaleEndLatitude, scaleEndLongitude))
  ];
  const distanceInPixels = Math.sqrt(
    ((points[0].x - points[1].x) * scale) ** degree + ((points[0].y - points[1].y) * scale) ** degree
  );
  return distanceInPixels / scaleDistance;
};

const degree = 2;

const canBeScaled = (scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude) => {
  return (scaleStartLatitude === 0 && scaleStartLongitude === 0) || (scaleEndLatitude === 0 && scaleEndLongitude === 0);
};

export const rotateImage = (map, google, level) => {
  const { scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude } = level;

  if (canBeScaled(scaleStartLatitude, scaleStartLongitude, scaleEndLatitude, scaleEndLongitude)) {
    return 0;
  }
  const A = map.getProjection().fromLatLngToPoint(new google.maps.LatLng(scaleStartLatitude, scaleStartLongitude));
  const B = map.getProjection().fromLatLngToPoint(new google.maps.LatLng(scaleEndLatitude, scaleEndLongitude));
  const hypotenuse = Math.sqrt((A.x - B.x) ** degree + (A.y - B.y) ** degree);
  const cathetus = Math.abs(A.y - B.y);
  let alpha = Math.acos(cathetus / hypotenuse);
  if ((A.x > B.x && A.y > B.y) || (A.x < B.x && A.y < B.y)) {
    alpha = alpha - Math.PI;
  }
  const minForAllRotate = 0;
  const maxForAllRotate = -1;
  if (minForAllRotate < alpha && alpha < maxForAllRotate) {
    return alpha;
  }
  return 0;
};
