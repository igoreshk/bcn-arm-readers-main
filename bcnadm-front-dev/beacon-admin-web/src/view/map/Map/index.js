import { hideLoadingScreen, showLoadingScreen } from 'src/actions/loadingScreen';
import { BuildingService } from 'src/service/BuildingService';
import React, { useEffect, useRef, useState } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { addPopupMessageText } from 'src/actions/popupMessage';
import { ERROR } from 'src/utils/popUpConsts';
import { BUILDINGS } from 'src/consts/RouteConsts';
import { withLocalize } from 'react-localize-redux';
import { setMapProvider, saveMap, setMapRenderedStatus } from 'src/reducers/mapSettingsReducer/mapSettingsSlice';
import { setCurrentBuilding, setCurrentLevel } from 'src/reducers/activeLevelInfoReducer/activeLevelInfoSlice';
import { LevelService } from 'service/LevelService';
import { setMapForGoogle } from 'view/map/DrawingBar/serviceForGoogleMaps';
import { INITIAL_ZOOM } from '../MapConsts';
import GoogleMapConfiguration from '../GoogleMapConfiguration';
import { calculatePixelsPerMeter, rotateImage, toggleBackgroundIfMapExist } from './helpers';
import './map.scss';

export const Map = (props) => {
  const currentOverlay = useRef(null);
  const [componentShouldUpdate, setComponentShouldUpdate] = useState(false);
  const mapContainer = useRef();
  const isMapRendered = useRef(false);
  const mapConfiguration = useRef();
  const currentLevelInfo = useRef(null);
  const currentBuildingInfo = useRef(null);

  const redirectTo = (path) => {
    props.hideLoadingScreen();
    props.history.push({
      pathname: path
    });
    props.addPopupMessageText(props.translate('building.messages.unavailable'), ERROR);
  };

  const getLevel = async () => {
    return await LevelService.getLevel(props.match.params.level);
  };

  const calculateSizeAndRotation = async (map, match) => {
    map.setZoom(INITIAL_ZOOM);
    const { google, temporaryScaleEdge } = props;
    if (!map.getProjection()) {
      props.google.maps.event.addListenerOnce(map, 'projection_changed', () => {
        calculateSizeAndRotation(map, match);
      });
      return;
    }
    const size = { height: currentBuildingInfo.current.height, width: currentBuildingInfo.current.width };

    map.panTo({ lat: currentBuildingInfo.current.latitude, lng: currentBuildingInfo.current.longitude });
    props.saveMap({ selectedMap: map });

    const pixelsPerMeter = await calculatePixelsPerMeter(map, google, {
      ...currentLevelInfo.current,
      ...temporaryScaleEdge
    });
    const rotation = await rotateImage(map, google, { ...currentLevelInfo.current, ...temporaryScaleEdge });
    await settingMap(map, match, size, pixelsPerMeter, rotation);
  };

  const initializeMap = async (buildingId) => {
    const { google, match, mapProvider, setBuilding } = props;
    props.setMapProvider({ selectedMapProvider: mapProvider });
    const node = mapContainer.current;

    try {
      const { setBuildingCenterCoordinates } = props;
      const building = await BuildingService.findOne(buildingId);
      await setBuilding({ building });
      mapConfiguration.current = await new GoogleMapConfiguration(google, node, {
        lat: building.latitude,
        lng: building.longitude
      });
      setBuildingCenterCoordinates([building.latitude, building.longitude]);
      const map = await mapConfiguration.current?.createMap();
      props.saveMap({ selectedMap: map });
      if (map.getProjection() && currentBuildingInfo.current !== null) {
        await calculateSizeAndRotation(map, match);
      }
    } catch (err) {
      console.error(err);
    }
  };

  const changeOverlay = (overlay, map) => {
    const { setPicture } = props;
    if (currentOverlay.current) {
      currentOverlay.current.setMap(null);
    }
    currentOverlay.current = overlay;
    currentOverlay.current.setMap(map);

    setPicture(currentOverlay.current.image);
  };

  const saveCoordinates = () => {
    const southWest = currentOverlay.current.bounds.getSouthWest();
    const northEast = currentOverlay.current.bounds.getNorthEast();
    if (
      currentLevelInfo.current.northEastLatitude !== northEast.lat() ||
      currentLevelInfo.current.northEastLongitude !== northEast.lng() ||
      currentLevelInfo.current.southWestLatitude !== southWest.lat() ||
      currentLevelInfo.current.southWestLongitude !== southWest.lng()
    ) {
      const level = currentLevelInfo.current;
      level.northEastLatitude = northEast.lat();
      level.northEastLongitude = northEast.lng();
      level.southWestLatitude = southWest.lat();
      level.southWestLongitude = southWest.lng();
      setCurrentLevel({ level });
      LevelService.saveLevel({ ...level });
    }
  };

  const toggleBackground = () => {
    const { selectedMap, showMapBackground } = props;
    mapConfiguration.current.setMapBackground(selectedMap, showMapBackground);
  };

  async function settingMap(map, match, size, pixelsPerMeter, rotation) {
    try {
      if (!mapConfiguration.current) {
        await initializeMap(props.match.params.building);
      }
      const { bounds, imageLink } = await mapConfiguration.current.getImageAndCalculateBounds(
        map,
        match.params.level,
        size,
        pixelsPerMeter
      ); // calculating bounds
      currentOverlay.current = await mapConfiguration.current.configureOverlay(map, bounds, imageLink, rotation); // configuring overlay
      const { setPictureCoords } = props;
      const coordinates = Object.values(currentOverlay.current.bounds);
      const data = coordinates.map(() => [
        { lat: coordinates[0].g, lng: coordinates[1].h },
        { lat: coordinates[0].h, lng: coordinates[1].g }
      ]);
      setPictureCoords(data);
      await changeOverlay(currentOverlay.current, map);
      saveCoordinates();
      props.hideLoadingScreen();
      props.setMapRenderedStatus({ isMapRendered: true });
    } catch (err) {
      console.error(err);
      redirectTo(BUILDINGS);
    }
  }

  useEffect(async () => {
    const buildingImage = currentOverlay.current?.image;
    await setMapForGoogle(props, buildingImage);
  }, [
    props.selectedMapProviderName,
    props.loadedOSMap,
    props.match,
    props.buildingCenterCoordinates,
    props.pictureCoords,
    props.translate,
    props.google,
    props.selectedMap,
    props.shapes
  ]);

  // Equivalent of ComponentDidMount
  useEffect(async () => {
    isMapRendered.current = false;
    props.showLoadingScreen();
    props.setMapRenderedStatus({ isMapRendered: false });
    currentLevelInfo.current = await getLevel();
    const buildingId = props.match.params.building;
    currentBuildingInfo.current = await BuildingService.findOne(buildingId);
    await initializeMap(buildingId);

    return () => {
      currentOverlay.current = null;
      currentLevelInfo.current = null;
    };
  }, []);

  useEffect(() => {
    setComponentShouldUpdate(true);
  }, [props.match.params.level]);

  useEffect(() => {
    if (currentBuildingInfo.current === null) {
      return;
    }

    if (props.loading && componentShouldUpdate && props.selectedMap) {
      props.showLoadingScreen();
      setComponentShouldUpdate(false);
      calculateSizeAndRotation(props.selectedMap, props.match).catch((err) => throw err);
    }
    if (mapConfiguration.current && !props.loading && props.selectedMap) {
      toggleBackgroundIfMapExist(props.selectedMap, toggleBackground);
    }
  }, [
    props.match.params.level,
    props.currentBuilding,
    props.selectedMap,
    currentLevelInfo.current,
    currentBuildingInfo.current
  ]);

  return (
    <div id="map" className="mapWrapper">
      <div ref={mapContainer} className="mapContainer" />
    </div>
  );
};

Map.propTypes = {
  google: PropTypes.object,
  mapProvider: PropTypes.object,
  showLoadingScreen: PropTypes.func.isRequired,
  setMapProvider: PropTypes.func.isRequired,
  saveMap: PropTypes.func.isRequired,
  setMapRenderedStatus: PropTypes.func,
  selectedMap: PropTypes.object,
  hideLoadingScreen: PropTypes.func.isRequired,
  match: PropTypes.object,
  showMapBackground: PropTypes.bool,
  addPopupMessageText: PropTypes.func,
  translate: PropTypes.func
};

const mapDispatchToProps = {
  showLoadingScreen,
  setBuilding: setCurrentBuilding,
  setMapProvider,
  saveMap,
  setMapRenderedStatus,
  hideLoadingScreen,
  addPopupMessageText
};

const mapStateToProps = (state) => {
  return {
    currentLevel: state.activeLevelInfo.currentLevel,
    currentBuilding: state.activeLevelInfo.currentBuilding,
    temporaryScaleEdge: state.activeLevelInfo.temporaryScaleEdge,
    selectedMap: state.mapSettings.selectedMap,
    loading: state.loading.processesFetching,
    showMapBackground: state.mapBackground.showMapBackground
  };
};
export default withRouter(withLocalize(connect(mapStateToProps, mapDispatchToProps)(Map)));
