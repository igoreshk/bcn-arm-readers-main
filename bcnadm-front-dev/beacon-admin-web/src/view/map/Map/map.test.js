import React from 'react';
import { mount } from 'enzyme';
import { INITIAL_ZOOM } from 'view/map/MapConsts';
import { calculatePixelsPerMeter, rotateImage } from './helpers';
import { Map } from './index';
import * as helpers from './helpers';

const getProjection = jest.fn(() => {
    return {
        fromLatLngToPoint: jest.fn().mockReturnValue({ x: 141, y: 70 }),
        fromPointToLatLng: jest.fn()
    };
});


const selectedMap = {
    setCenter: jest.fn(),
    getCenter: jest.fn().mockReturnValue({ x: 128, y: 128 }),
    getZoom: jest.fn().mockReturnValue(INITIAL_ZOOM),
    getProjection,
    setZoom: jest.fn(),
    panTo: jest.fn()
};

class LatLng {
    Ta = {
        g: -0.3785,
        i: 0.3785
    };
    Ya = {
        g: -0.3785,
        i: 0.3785
    };

    extend() {
        jest.fn();
    }
}

class LatLngBounds {
    extend() {
        jest.fn();
    }
}

const google = {
    maps: {
        Map: jest.fn(() => selectedMap),
        OverlayView: jest.fn(),
        LatLngBounds,
        LatLng,
        Point: jest.fn().mockReturnValue({ x: 128, y: 128 }),
        event: {
            addListenerOnce: jest.fn()
        }
    }
};

const match = {
    isExact: true,
    params: {
        building: '5dd3c1d2f79fb2000125f0c0',
        edit: true,
        layer: 'beacons',
        level: '5dd3c219f79fb2000125f0c1'
    },
    path: '/buildings/:building/levels/:level/:layer/:edit?',
    url: '/buildings/5dd3c1d2f79fb2000125f0c0/levels/5dd3c219f79fb2000125f0c1/beacons'
};
const coordinate = {
    latitude: 50,
    longitude: 50
};
const size = {
    width: 50,
    height: 50
};

const currentLevel = {
    scaleStartLatitude: 59.888661046666236,
    scaleStartLongitude: 30.326315760612488,
    scaleEndLatitude: 59.88888651067054,
    scaleEndLongitude: 30.328509944346113,
    scaleDistance: 0
};

const level = {
    scaleStartLatitude: 59.888661046666236,
    scaleStartLongitude: 30.326315760612488,
    scaleEndLatitude: 59.88888651067054,
    scaleEndLongitude: 30.328509944346113,
    scaleDistance: 40
};

const curLevel = {
    scaleStartLatitude: 0,
    scaleStartLongitude: 0,
    scaleEndLatitude: 0,
    scaleEndLongitude: 0,
    scaleDistance: 40
};

const currentBuilding = {
    height: 150,
    width: 200
};

const props = {
    google,
    hideLoadingScreen: jest.fn(),
    showLoadingScreen: jest.fn(),
    setMapRenderedStatus: jest.fn(),
    setCurrentBuilding: jest.fn(),
    setMapProvider: jest.fn(),
    match,
    saveGoogle: jest.fn(),
    saveMap: jest.fn(),
    history: {
        push: jest.fn()
    },
    location: {
        hash: '',
        key: 'zjzhge',
        pathname: '/buildings/5dd3c1d2f79fb2000125f0c0/levels/5dd3c219f79fb2000125f0c1/beacons',
        search: ''
    },
    selectedMap,
    showMapBackground: true,
    staticContext: undefined,
    coordinate,
    loading: false,
    size,
    currentLevel,
    currentBuilding
};

jest.mock('service/BuildingService');


describe('Map', () => {
    let wrapper;
    let container;
    jest.spyOn(helpers, 'toggleBackgroundIfMapExist').mockImplementation();
    const waitForPromises = () => new Promise(setImmediate);

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = mount(<Map {...props} />);
        container = wrapper.find(Map);
    });

    it('Checks that props are passed properly', () => {
        expect(container.props()).toEqual(props);
    });

    it('Checks showLoadingScreen have been called', async () => {
        const showLoadingScreen = jest.spyOn(container.props(), 'showLoadingScreen');
        await waitForPromises();
        // eslint-disable-next-line require-atomic-updates
        container.mapConfiguration = {
            getImageAndCalculateBounds: jest.fn(() => Promise.resolve({ bounds: {}, imageLink: '' }))
        };
        wrapper.setProps({ match: { params: { level: '' } } });
        expect(showLoadingScreen).toHaveBeenCalled();
    });

    it('Check matching snapshot', () => expect(wrapper.html()).toMatchSnapshot());

    it('returns baseValue if distance equals to zero', () => {
        expect(calculatePixelsPerMeter(selectedMap, google, { ...currentLevel })).toEqual(10);
    });

    it('returns calculated value if distance not equals to zero', () => {
        expect(calculatePixelsPerMeter(selectedMap, google, { ...level })).toEqual(0);
    });
    it('returns 0 if level params equals to zero', () => {
        expect(rotateImage(selectedMap, google, { ...curLevel })).toEqual(0);
    });

    it('returns calculated value if distance not equals to zero', () => {
        expect(calculatePixelsPerMeter(selectedMap, google, { ...level })).toEqual(0);
    });
    it('returns 0 if level params equals to zero', () => {
        expect(rotateImage(selectedMap, google, { ...curLevel })).toEqual(0);
    });
});
