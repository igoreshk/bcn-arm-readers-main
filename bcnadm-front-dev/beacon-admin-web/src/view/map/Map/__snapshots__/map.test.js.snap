// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Check matching snapshot 1`] = `"<div id=\\"map\\" class=\\"mapWrapper\\"><div class=\\"mapContainer\\"></div></div>"`;
